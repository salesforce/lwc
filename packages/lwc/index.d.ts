/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

/**
 * Lightning Web Components core module
 */

declare module 'lwc' {
    // backwards compatible type used for the old days when TS didn't support `event.composed`
    interface ComposableEvent extends Event {
        composed: boolean;
    }

    class HTMLElementTheGoodPart {
        dispatchEvent(evt: Event): boolean;
        addEventListener(
            type: string,
            listener: EventListenerOrEventListenerObject,
            options?: boolean | AddEventListenerOptions
        ): void;
        removeEventListener(
            type: string,
            listener: EventListenerOrEventListenerObject,
            options?: boolean | EventListenerOptions
        ): void;
        setAttributeNS(ns: string, attrName: string, value: any): void;
        removeAttributeNS(ns: string, attrName: string): void;
        removeAttribute(attrName: string): void;
        setAttribute(attrName: string, value: any): void;
        getAttribute(attrName: string): string | null;
        getAttributeNS(ns: string, attrName: string): string | null;
        getBoundingClientRect(): ClientRect;
        querySelector<E extends Element = Element>(selectors: string): E | null;
        querySelectorAll<E extends Element = Element>(selectors: string): NodeListOf<E>;
        getElementsByTagName(tagNameOrWildCard: string): HTMLCollectionOf<Element>;
        getElementsByClassName(names: string): HTMLCollectionOf<Element>;
        attachInternals(): ElementInternals;
        readonly tagName: string;
        readonly classList: DOMTokenList;

        // Default HTML Properties
        dir: string;
        id: string;
        accessKey: string;
        title: string;
        lang: string;
        hidden: boolean;
        draggable: boolean;
        tabIndex: number;

        // Aria Properties
        ariaAutoComplete: string | null;
        ariaChecked: string | null;
        ariaCurrent: string | null;
        ariaDisabled: string | null;
        ariaExpanded: string | null;
        ariaHasPopup: string | null;
        ariaHidden: string | null;
        ariaInvalid: string | null;
        ariaLabel: string | null;
        ariaLevel: string | null;
        ariaMultiLine: string | null;
        ariaMultiSelectable: string | null;
        ariaOrientation: string | null;
        ariaPressed: string | null;
        ariaReadOnly: string | null;
        ariaRequired: string | null;
        ariaSelected: string | null;
        ariaSort: string | null;
        ariaValueMax: string | null;
        ariaValueMin: string | null;
        ariaValueNow: string | null;
        ariaValueText: string | null;
        ariaLive: string | null;
        ariaRelevant: string | null;
        ariaAtomic: string | null;
        ariaBusy: string | null;
        ariaActiveDescendant: string | null;
        ariaControls: string | null;
        ariaDescribedBy: string | null;
        ariaFlowTo: string | null;
        ariaLabelledBy: string | null;
        ariaOwns: string | null;
        ariaPosInSet: string | null;
        ariaSetSize: string | null;
        ariaColCount: string | null;
        ariaColIndex: string | null;
        ariaColIndexText: string | null;
        ariaDescription: string | null;
        ariaDetails: string | null;
        ariaErrorMessage: string | null;
        ariaKeyShortcuts: string | null;
        ariaModal: string | null;
        ariaPlaceholder: string | null;
        ariaRoleDescription: string | null;
        ariaRowCount: string | null;
        ariaRowIndex: string | null;
        ariaRowIndexText: string | null;
        ariaRowSpan: string | null;
        ariaColSpan: string | null;
        ariaBrailleLabel: string | null;
        ariaBrailleRoleDescription: string | null;
        role: string | null;
    }

    // @ts-ignore type-mismatch
    interface ShadowRootTheGoodPart extends NodeSelector {
        mode: string;
        readonly activeElement: Element | null;
        readonly host: HTMLElement;
        readonly firstChild: Node | null;
        readonly lastChild: Node | null;
        readonly innerHTML: string;
        readonly textContent: string;
        readonly childNodes: Node[];
        readonly delegatesFocus: boolean;
        addEventListener(
            type: string,
            listener: EventListenerOrEventListenerObject,
            options?: boolean | AddEventListenerOptions
        ): void;
        removeEventListener(
            type: string,
            listener: EventListenerOrEventListenerObject,
            options?: boolean | EventListenerOptions
        ): void;
        hasChildNodes(): boolean;
        compareDocumentPosition(otherNode: Node): number;
        contains(otherNode: Node): boolean;
        elementFromPoint(x: number, y: number): Element | null;
        querySelector<K extends keyof HTMLElementTagNameMap>(
            selectors: K
        ): HTMLElementTagNameMap[K] | null;
        querySelector<K extends keyof SVGElementTagNameMap>(
            selectors: K
        ): SVGElementTagNameMap[K] | null;
        querySelector<E extends Element = Element>(selectors: string): E | null;
        querySelectorAll<K extends keyof HTMLElementTagNameMap>(
            selectors: K
        ): NodeListOf<HTMLElementTagNameMap[K]>;
        querySelectorAll<K extends keyof SVGElementTagNameMap>(
            selectors: K
        ): NodeListOf<SVGElementTagNameMap[K]>;
        querySelectorAll<E extends Element = Element>(selectors: string): NodeListOf<E>;
    }

    /**
     * Base class for the Lightning Web Component JavaScript class
     */
    export class LightningElement extends HTMLElementTheGoodPart {
        /**
         * This static getter builds a Web Component class from a LWC constructor so it can be registered
         * as a new element via customElements.define() at any given time. For example:
         *
         * ```
         * import XComponent from 'namespace/element';
         * customElements.define('x-component', XComponent.CustomElementConstructor);
         * const elm = document.createElement('x-component');
         * ```
         */
        static get CustomElementConstructor(): typeof HTMLElement;
        /**
         * Called when the element is inserted in a document
         */
        connectedCallback(): void;
        /**
         * Called when the element is removed from a document
         */
        disconnectedCallback(): void;
        /**
         * Called after every render of the component
         */
        renderedCallback(): void;
        /**
         * Called when a descendant component throws an error in one of its lifecycle hooks
         */
        errorCallback(error: Error, stack: string): void;

        readonly template: ShadowRootTheGoodPart;
        readonly shadowRoot: null;
        readonly refs: { [key: string]: Element | LightningElement };
    }

    /**
     * Decorator to mark public reactive properties
     */
    export const api: PropertyDecorator;

    /**
     * Decorator to mark private reactive properties
     */
    export const track: PropertyDecorator;

    type StringKeyedRecord = Record<string, any>;
    /**
     * Decorator factory to wire a property or method to a wire adapter data source
     * Use generic types to allow type checking for wire adapters
     * Default all the generic types to any to maintain backward compatibility
     *
     * For example, a wire adapter 'getRecord' can have the following type definition
     *
     * export const getRecord: WireAdapterConstructor<GetRecordConfig, RecordRepresentation>;
     *
     * in which 'GetRecordConfig' is the adapter config object type and 'RecordRepresentation'
     * is the returned value.
     *
     * @param adapter the adapter used to provision data
     * @param config configuration object for the adapter
     */
    export function wire<
        Config extends StringKeyedRecord,
        Value,
        Context extends StringKeyedRecord = StringKeyedRecord
    >(
        adapter:
            | WireAdapterConstructor<Config, Value, Context>
            | LegacyWireAdapterConstructor<Config, Value>,
        config?: WireConfigValue<Config>
    ): PropertyDecorator;

    type LegacyWireAdapterConstructor<Config, Value> = (config?: Config) => Value;
    type WireConfigValue<Config extends StringKeyedRecord = StringKeyedRecord> = {
        // wire reactive variables are strings prefixed with '$' so the config value can just be string
        [K in keyof Config]: Config[K] | string;
    };
    type ContextValue<Context extends StringKeyedRecord = StringKeyedRecord> = Context;

    interface WireAdapter<Config extends StringKeyedRecord, Context extends StringKeyedRecord> {
        update(config: WireConfigValue<Config>, context?: ContextValue<Context>): void;
        connect(): void;
        disconnect(): void;
    }

    type WireDataCallback<Value> = (value: Value) => void;
    type WireAdapterSchemaValue = 'optional' | 'required';

    interface ContextConsumer<Context extends StringKeyedRecord> {
        provide(newContext: ContextValue<Context>): void;
    }

    interface ContextProviderOptions<Context extends StringKeyedRecord> {
        consumerConnectedCallback: (consumer: ContextConsumer<Context>) => void;
        consumerDisconnectedCallback?: (consumer: ContextConsumer<Context>) => void;
    }

    interface WireAdapterConstructor<
        Config extends StringKeyedRecord,
        Value,
        Context extends StringKeyedRecord
    > {
        new (callback: WireDataCallback<Value>): WireAdapter<Config, Context>;
        configSchema?: Record<keyof Config, WireAdapterSchemaValue>;
        contextSchema?: Record<keyof Context, WireAdapterSchemaValue>;
    }

    type Contextualizer<Context extends StringKeyedRecord> = (
        elm: EventTarget,
        options: ContextProviderOptions<Context>
    ) => void;
    export function createContextProvider<
        Config extends StringKeyedRecord,
        Value,
        Context extends StringKeyedRecord = StringKeyedRecord
    >(config: WireAdapterConstructor<Config, Value, Context>): Contextualizer<Context>;
}
