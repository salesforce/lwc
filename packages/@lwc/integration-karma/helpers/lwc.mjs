import path from 'node:path';
import { rollup } from 'rollup';
import lwcRollupPlugin from '@lwc/rollup-plugin';

import {
    DISABLE_SYNTHETIC_SHADOW_SUPPORT_IN_COMPILER,
    API_VERSION,
    DISABLE_STATIC_CONTENT_OPTIMIZATION,
} from './options.mjs';

// Cache reused between each compilation to speed up the compilation time.
let cache;

export default async (ctx) => {
    const input = ctx.path.slice(1); // strip leading / from URL path to get relative file path
    const suiteDir = path.dirname(input);

    // TODO [#3370]: remove experimental template expression flag
    const experimentalComplexExpressions = suiteDir.includes('template-expressions');

    const createRollupPlugin = (options) => {
        return lwcRollupPlugin({
            // Sourcemaps don't work with Istanbul coverage
            sourcemap: !process.env.COVERAGE,
            experimentalDynamicComponent: {
                loader: 'test-utils',
                strict: true,
            },
            enableDynamicComponents: true,
            enableLwcOn: true,
            experimentalComplexExpressions,
            enableStaticContentOptimization: !DISABLE_STATIC_CONTENT_OPTIMIZATION,
            disableSyntheticShadowSupport: DISABLE_SYNTHETIC_SHADOW_SUPPORT_IN_COMPILER,
            apiVersion: API_VERSION,
            ...options,
        });
    };

    const defaultRollupPlugin = createRollupPlugin();

    const customLwcRollupPlugin = {
        ...defaultRollupPlugin,
        transform(src, id) {
            let rollupPluginToUse;

            // Override the LWC Rollup plugin to specify different options based on file name patterns.
            // This allows us to alter the API version or other compiler props on a filename-only basis.
            const apiVersion = id.match(/useApiVersion(\d+)/)?.[1];
            const nativeOnly = /\.native-only\./.test(id);
            if (apiVersion) {
                rollupPluginToUse = createRollupPlugin({
                    apiVersion: parseInt(apiVersion, 10),
                });
            } else if (nativeOnly) {
                rollupPluginToUse = createRollupPlugin({ disableSyntheticShadowSupport: true });
            } else {
                rollupPluginToUse = defaultRollupPlugin;
            }
            return rollupPluginToUse.transform.call(this, src, id);
        },
    };

    const plugins = [customLwcRollupPlugin];

    const bundle = await rollup({
        input,
        plugins,
        cache,

        // Rollup should not attempt to resolve the engine and the test utils, Karma takes care of injecting it
        // globally in the page before running the tests.
        external: ['lwc', 'wire-service', 'test-utils', '@test/loader'],

        onwarn(warning, warn) {
            // Ignore warnings from our own Rollup plugin
            if (warning.plugin !== 'rollup-plugin-lwc-compiler') {
                warn(warning);
            }
        },
    });

    cache = bundle.cache;

    const { output } = await bundle.generate({
        format: 'esm',
        // TODO: Does web-test-runner use istanbul?
        // Sourcemaps don't work with Istanbul coverage
        sourcemap: process.env.COVERAGE ? false : 'inline',

        // The engine and the test-utils is injected as UMD. This mapping defines how those modules can be
        // referenced from the window object.
        globals: {
            lwc: 'LWC',
            'wire-service': 'WireService',
            'test-utils': 'TestUtils',
        },
    });

    const { code } = output[0];
    return code;
};
