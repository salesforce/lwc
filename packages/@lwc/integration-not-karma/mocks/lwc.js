// IMPORTANT: we must use @lwc/engine-dom instead of lwc in order to avoid circular imports
import { sanitizeAttribute as _sanitizeAttribute } from '@lwc/engine-dom';
import { fn } from '@vitest/spy';

export * from '@lwc/engine-dom';
export const sanitizeAttribute = fn(_sanitizeAttribute);
