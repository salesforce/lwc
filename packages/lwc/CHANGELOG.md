# Changelog

## [7.2.3](https://github.com/salesforce/lwc/compare/v7.2.2...v7.2.3) (2024-08-30)


### Dependencies

* The following workspace dependencies were updated
  * dependencies
    * @lwc/compiler bumped from 7.2.2 to 7.2.3
    * @lwc/engine-server bumped from 7.2.2 to 7.2.3
    * @lwc/rollup-plugin bumped from 7.2.2 to 7.2.3
    * @lwc/template-compiler bumped from 7.2.2 to 7.2.3
