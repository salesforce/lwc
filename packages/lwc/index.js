/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const fs = require('fs');
const path = require('path');

const DEBUG_SUFFIX = '_debug';
const PROD_SUFFIX = '.min';

function getVersion() {
    const pkgPath = path.join(__dirname, './package.json');
    const pkgSrc = fs.readFileSync(pkgPath, 'utf8');
    const pkg = JSON.parse(pkgSrc);
    return pkg.version;
}

const MODULES = new Set(['engine-dom', 'engine-server', 'synthetic-shadow', 'wire-service']);
const FORMATS = new Set(['iife', 'umd', 'esm']);
const TARGETS = new Set(['es5', 'es2017']);
const MODES = new Set(['dev', 'prod', 'prod_debug']);

const DEFAULT_FORMAT = 'esm';
const DEFAULT_TARGET = 'es2017';
const DEFAULT_MODE = 'dev';

function validateArgs(name, format, target, mode) {
    if (!MODULES.has(name)) {
        throw new Error(
            `Invalid module name "${name}". Available module names: ${Array.from(MODULES.keys())}`
        );
    }

    if (!FORMATS.has(format)) {
        throw new Error(
            `Invalid format "${name}". Available formats: ${Array.from(FORMATS.keys())}`
        );
    }

    if (!TARGETS.has(target)) {
        throw new Error(
            `Invalid target "${target}". Available targets: ${Array.from(TARGETS.keys())}`
        );
    }

    if (!MODES.has(mode)) {
        throw new Error(`Invalid mode "${mode}". Available modes: ${Array.from(MODES.keys())}`);
    }
}

function getModule(name, format = DEFAULT_FORMAT, target = DEFAULT_TARGET, mode = DEFAULT_MODE) {
    const distPath = getModulePath(name, format, target, mode);
    return fs.readFileSync(distPath, 'utf8');
}

function getModulePath(
    name,
    format = DEFAULT_FORMAT,
    target = DEFAULT_TARGET,
    mode = DEFAULT_MODE
) {
    validateArgs(name, format, target, mode);
    let distPath;
    // The default is on the package itself (naming convention is set)
    if (format === DEFAULT_FORMAT && target === DEFAULT_TARGET && mode === DEFAULT_MODE) {
        distPath = require.resolve(`@lwc/${name}/dist/${name}.js`);

        // Otherwise is on dist of this package
    } else if (mode === 'prod') {
        distPath = path.join(__dirname, 'dist', name, format, target, `${name}${PROD_SUFFIX}.js`);
    } else if (mode === 'prod_debug') {
        distPath = path.join(__dirname, 'dist', name, format, target, `${name}${DEBUG_SUFFIX}.js`);
    } else {
        distPath = path.join(__dirname, 'dist', name, format, target, `${name}.js`);
    }

    if (!fs.existsSync(distPath)) {
        throw new Error(`Module path "${distPath}" for module ${name} not found`);
    }

    return distPath;
}

module.exports = {
    getVersion,
    getModule,
    getModulePath,
};
