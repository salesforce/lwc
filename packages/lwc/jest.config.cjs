/*
 * Copyright (c) 2024, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/* eslint-env node */

/**
 * This is a .cjs file because this package is an ESM package, but jest needs CJS to work.
 * If this file is ever changed to a .js file, add an exclusion to the package.json "files" list,
 * so that this is not shipped as part of the package.
 */

// eslint-disable-next-line @typescript-eslint/no-var-requires
const BASE_CONFIG = require('../../scripts/jest/base.config');

module.exports = {
    ...BASE_CONFIG,
    displayName: 'lwc',
    testEnvironment: 'jsdom',
};
