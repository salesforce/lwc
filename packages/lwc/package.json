{
    "name": "lwc",
    "version": "2.17.0",
    "description": "Lightning Web Components (LWC)",
    "homepage": "https://lwc.dev/",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/lwc.git",
        "directory": "packages/lwc"
    },
    "bugs": {
        "url": "https://github.com/salesforce/lwc/issues"
    },
    "main": "index.js",
    "module": "dist/engine-dom/esm/es2017/engine-dom.js",
    "typings": "types.d.ts",
    "license": "MIT",
    "scripts": {
        "clean": "rm -rf dist/",
        "build": "node scripts/build.js",
        "dev": "node scripts/build.js --watch"
    },
    "files": [
        "dist/",
        "index.js",
        "types.d.ts"
    ],
    "lwc": {
        "modules": [
            {
                "name": "lwc",
                "path": "dist/engine-dom/esm/es2017/engine-dom.js"
            },
            {
                "name": "@lwc/synthetic-shadow",
                "path": "dist/synthetic-shadow/esm/es2017/synthetic-shadow.js"
            },
            {
                "name": "@lwc/wire-service",
                "path": "dist/synthetic-shadow/esm/es2017/wire-service.js"
            }
        ]
    },
    "dependencies": {
        "@lwc/compiler": "2.17.0",
        "@lwc/engine-dom": "2.17.0",
        "@lwc/engine-server": "2.17.0",
        "@lwc/features": "2.17.0",
        "@lwc/synthetic-shadow": "2.17.0",
        "@lwc/wire-service": "2.17.0"
    }
}
