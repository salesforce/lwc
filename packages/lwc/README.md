Lightning Web Components (LWC) is an enterprise-grade web components foundation for building user interfaces. LWC provides a simple authoring format for UI Components. LWC compiles this authoring format into low-level Web Component APIs. The `lwc` package is the main entry point for dependencies.

-   Develop components quickly and declaratively using HTML, JavaScript, and CSS.
-   Develop accessible components so that everyone can understand and navigate your app.
-   Components are encapsulated in all browsers via the `@lwc/synthetic-shadow` package, which polyfills the Shadow DOM.

Developing a Lightning web component is this easy.

```ascii
counter
  ├──counter.css
  ├──counter.html
  └──counter.js
```

```html
<!-- counter.html -->
<template>
    <p>Counter: {count}</p>
    <button onclick="{increaseCounter}">Add</button>
</template>
```

```css
/* counter.css */
p {
    font-family: serif;
    font-size: large;
}
```

```javascript
// counter.js
import { LightningElement } from 'lwc';

export default class Counter extends LightningElement {
    count = 0;

    increaseCounter() {
        this.count += 1;
    }
}
```

## Supported Browsers

| Browser                       | Version |
| ----------------------------- | ------- |
| Microsoft® Internet Explorer® | IE 11\* |
| Microsoft® Edge               | Latest  |
| Google Chrome™                | Latest  |
| Mozilla® Firefox®             | Latest  |
| Apple® Safari®                | 12.x+   |

For IE 11, LWC uses compatibility mode. Code is transpiled down to ES5 and the required polyfills are added. To develop components that run in IE 11, follow the [compatibility mode](https://github.com/salesforce/eslint-plugin-lwc#compat-performance) linting rules.

## Docs, Recipes, & Support

[lwc.dev](https://lwc.dev) has all the information you need to develop components using LWC, including [code recipes](https://recipes.lwc.dev/) and code playgrounds.

Get started fast using the [`create-lwc-app`](https://www.npmjs.com/package/create-lwc-app) tool.

```bash
npx create-lwc-app my-app
cd my-app
npm run watch
```

For support, use the `lwc` tag on [Stack Overflow](https://stackoverflow.com/questions/tagged/lwc) or the `lightning-web-components` tag on [Salesforce Stack Exchange](https://salesforce.stackexchange.com/questions/tagged/lightning-web-components).

## Release Notes

Changes are documented at [github.com/salesforce/lwc/releases](https://github.com/salesforce/lwc/releases).
