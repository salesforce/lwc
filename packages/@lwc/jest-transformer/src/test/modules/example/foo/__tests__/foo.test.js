/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
import { createElement } from 'lwc';
import Foo from 'example/foo';
import Bar from 'other/bar';

jest.mock('globalLib', () => {
    return {
        func: () => 'from foo.test.js',
    };
});

describe('example-foo', () => {
    afterEach(() => {
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
    });

    it('should render correct snapshot', () => {
        const element = createElement('example-foo', { is: Foo });
        document.body.appendChild(element);
        expect(element).toMatchSnapshot();
    });

    it('3renders foo-inner DOM', () => {
        const element = createElement('example-foo', { is: Foo });
        document.body.appendChild(element);
        const classes = element.shadowRoot
            .querySelector('example-foo-inner')
            .shadowRoot.querySelectorAll('.exampleFooInner');
        expect(classes).toHaveLength(1);
    });

    it('renders component from another namespace (other-bar)', () => {
        const element = createElement('example-foo', { is: Foo });
        document.body.appendChild(element);
        const classes = element.shadowRoot
            .querySelector('other-bar')
            .shadowRoot.querySelectorAll('.otherBar');
        expect(classes).toHaveLength(1);
    });

    it('can import component from another namespace', () => {
        const element = createElement('other-bar', { is: Bar });
        document.body.appendChild(element);
        const classes = element.shadowRoot.querySelectorAll('.otherBar');
        expect(classes).toHaveLength(1);
    });

    it('contains default property <me>', () => {
        const element = createElement('example-foo', { is: Foo });
        expect(element.me).toBe('foo');
    });

    it('can use async syntax', () => {
        const expected = { foo: 'bar' };
        const element = createElement('example-foo', { is: Foo });

        return element.asyncMethod(expected).then(ret => expect(ret).toBe(expected));
    });

    it('returns value on inner element from mock defined in test file', () => {
        const element = createElement('example-foo', { is: Foo });
        document.body.appendChild(element);
        const inner = element.shadowRoot.querySelector('example-foo-inner');
        expect(inner.globalLibReturn).toBe('from foo.test.js');
    });
});
