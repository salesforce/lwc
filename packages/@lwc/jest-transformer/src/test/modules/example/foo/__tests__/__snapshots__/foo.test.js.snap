// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example-foo should render correct snapshot 1`] = `
<example-foo>
  #shadow-root(open)
    <p>
      example-foo
    </p>
    <div
      class="fooClass"
    >
      Hi from 
      foo
    </div>
    <example-foo-inner>
      #shadow-root(open)
        <p
          class="exampleFooInner"
        >
          example-foo-inner
        </p>
        <slot />
      <div>
        Passed Element
      </div>
    </example-foo-inner>
    <other-bar>
      #shadow-root(open)
        <p
          class="otherBar"
        >
          other-bar
        </p>
    </other-bar>
</example-foo>
`;
